### DISCUSSION TITLE
Unknown arg `provider_providerAccountId` in where.provider_providerAccountId for type accountWhereUniqueInput. Available args:

### DISCUSSION THREAD

**Discussion by HT-Moh** [nextauthjs/next-auth#4653] [Help] [answered]
### Environment

System:
    OS: Linux 5.13 Ubuntu 20.04.4 LTS (Focal Fossa)
    CPU: (8) x64 Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
    Memory: 11.93 GB / 30.99 GB
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 16.14.2 - /usr/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 8.1.2 - /usr/local/bin/npm
  Browsers:
    Brave Browser: 100.1.37.116
    Chrome: 100.0.4896.127
    Firefox: 99.0
  npmPackages:
    next: 12.1.5 => 12.1.5 
    next-auth: ^4.3.3 => 4.3.3 
    react: ^18.0.0 => 18.0.0 

### Reproduction URL

https://github.com/nextauthjs/next-auth/discussions/3631

### Describe the issue

I upgraded and I followed https://next-auth.js.org/getting-started/upgrade-v4#database-migration but now I am getting the following when `synchronize = true` 

![image](https://user-images.githubusercontent.com/34759569/165356508-aed54fc5-11b3-45cc-9e79-4016e126021c.png)

and the following when  `synchronize:  true` :

![image](https://user-images.githubusercontent.com/34759569/165352517-98ed77c0-ab7b-40c7-b904-4593c84c3d4b.png)

 ` ` `
    "@next-auth/typeorm-legacy-adapter": "^0.1.2-next.20",
    "next": "12.1.5",
    "next-auth": "^4.3.3",
 ` ` ` 

I went back `import { PrismaAdapter } from "@next-auth/prisma-adapter"` 
same issue. 

![image](https://user-images.githubusercontent.com/34759569/165362894-e5459adb-1584-4a49-a416-331da51f490a.png)

I made sure that the schema as per 2 guides I follow but no. 
first one: https://next-auth.js.org/getting-started/upgrade-v4#schema-changes
second one: https://next-auth.js.org/getting-started/upgrade-v4#database-migration

Really confusing which adapter to choose and even what is the write schema. 

I had the same problem a while back and I made everything work and now after upgrading the app naturally everything stopped working. 

I have 2 simple questions:
Which adapter needs to be used `TypeORMLegacyAdapter` or `'@next-auth/prisma-adapter'` ?
What is the correct schema? 

### How to reproduce

here is the final schema, but I tested all possible schema I found 
```
model account {
  id                Int      @id(map: "accounts_pkey") @default(autoincrement())
  userId            Int
  type              String
  provider          String
  providerAccountId String
  refresh_token     String?
  access_token      String?
  expires_at        String?
  created_at        DateTime @default(now()) @db.Timestamptz(6)
  updated_at        DateTime @default(now()) @db.Timestamptz(6)
  session_state     String?
  id_token          String?
  scope             String?
  token_type        String?
  user              user     @relation(fields: [userId], references: [id], onUpdate: Restrict)
}
model sessions {
  id           Int      @id @default(autoincrement())
  expires      String
  sessionToken String
  created_at   DateTime @default(now()) @db.Timestamptz(6)
  updated_at   DateTime @default(now()) @db.Timestamptz(6)
  userId       Int
  user         user     @relation(fields: [userId], references: [id], onDelete: NoAction, onUpdate: NoAction, map: "fk_user_id")
}
model user {
  id                Int                 @id(map: "users_pkey") @default(autoincrement())
  name              String?
  email             String              @unique(map: "users_email_key")
  emailVerified     String?
  image             String?
  created_at        DateTime            @default(now()) @db.Timestamptz(6)
  updated_at        DateTime            @default(now()) @db.Timestamptz(6)
  is_admin          Boolean             @default(false)
  twitter           String?
  avatar            String?
  title             String?
  first_name        String?
  last_name         String?
  facebook          String?
  github            String?
  project_id        Int?
  projects          projects?           @relation(fields: [project_id], references: [id], onDelete: Restrict, onUpdate: Restrict, map: "users_project_id_fkey")
  account           account[]
  blog_author       blog_author[]
  comment           comment[]
  minter_collection minter_collection[]
  sessions          sessions[]
}

model verification_tokens {
  id         Int      @id(map: "verification_requests_pkey") @default(autoincrement())
  identifier String
  token      String   @unique(map: "verificationtokens_token_key")
  expires    String
  created_at DateTime @default(now()) @db.Timestamptz(6)
  updated_at DateTime @default(now()) @db.Timestamptz(6)
}

````

### Expected behavior

it should just work

**Comment by lucasnribeiro** [comment]
Did you found a solution?

**Comment by HT-Moh** [comment, marked as answer]
Actually, almost every time I updated the library I have such issues, the fields change often and their names do not follow the naming convention for DB, another issue is the last version of `prisma`, `npx prisma db pull` cmd seems to override the `@map` so I needed to update the DB to make everything works 
Here is what works for me on `next-auth 4.15.0` `prisma 4.5.0` `@next-auth/prisma-adapter 1.0.5`
change table name `accounts` to `account`
change table name `users` to `user`
change attribute on `provider_account_id` to `providerAccountId` 
change FK  `user_id` to `userId`  (FK on `account` table)
change attribute  `email_verified` to `emailVerified` on `user` table
Make `provider` and `providerAccountId` unique
On Hasura if you have to delete a unique make sure the index is deleted as well otherwise it prevents you from setting unique again.
last steps `npx prisma db pull` && `npx prisma generate`

**Reply by matsyui** [reply]
2024, For other people...

If this solution didn't helped you (like me). Hope my solution can help you:

**[Unknown argument provider_providerAccountId. Available options are marked with ?](https://github.com/nextauthjs/next-auth/issues/11205#issuecomment-2178251882)**

**Comment by teddybee** [comment]
I have the same issue with @next-auth/prisma-adapter": "^1.0.7, "@prisma/client": "^4.15.0" and sqlite.

Unknown arg `provider_providerAccountId` in where.provider_providerAccountId for type AccountWhereUniqueInput. Did you mean `providerId_providerAccountId

**Comment by hyoloui** [comment]
```
../node_modules/@next-auth/prisma-adapter/dist/index.js:211:45

  208 getUserByEmail: (email) => p.user.findUnique({ where: { email } }),
  209 async getUserByAccount(providerAccountId) {
  210     var _a;
→ 211     const account = await p.account.findUnique({
            where: {
              providerAccountId: {
              ~~~~~~~~~~~~~~~~~
                providerAccountId: '11111111',
                provider: 'github'
```


Changing the schema will still cause the problem.

**Comment by ManishBisht777** [comment]
```
model Account {
    id                 String    @id @default(cuid())
    userId             String
    providerType       String
    providerId         String
    role               String
    providerAccountId  String
    refreshToken       String?
    accessToken        String?
    accessTokenExpires DateTime?
    createdAt          DateTime  @default(now())
    updatedAt          DateTime  @updatedAt
    user               User      @relation(fields: [userId], references: [id])

    @@unique([providerId, providerAccountId])
    @@index([userId])
}

model Session {
    id           String   @id @default(cuid())
    userId       String
    expires      DateTime
    sessionToken String   @unique
    accessToken  String   @unique
    role         String
    createdAt    DateTime @default(now())
    updatedAt    DateTime @updatedAt
    user         User     @relation(fields: [userId], references: [id])

    @@index([userId])
}

model User {
    id            String    @id @default(cuid())
    name          String?
    email         String?   @unique
    emailVerified DateTime?
    image         String?
    role          String
    createdAt     DateTime  @default(now())
    updatedAt     DateTime  @updatedAt
    accounts      Account[]
    sessions      Session[]
}

model VerificationRequest {
    id         String   @id @default(cuid())
    identifier String
    token      String   @unique
    expires    DateTime
    createdAt  DateTime @default(now())
    updatedAt  DateTime @updatedAt

    @@unique([identifier, token])
}
```


This is my Prisma schema and i am still getting the 

Unknown argument `provider_providerAccountId`. Did you mean `providerId_providerAccountId`? Available options are listed in green. PrismaClientValidationError:
Invalid `prisma.account.findUnique()` invocation:

```
{
  where: {
    provider_providerAccountId: {
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
      providerAccountId: "113802192673870361114",
      provider: "google"
    },
?   id?: String,
...
```

**Reply by 777777Z77777777** [reply]
Did you solve it?

**Comment by hyoloui** [comment]
No…hahaha

**Comment by nicoalbanese** [comment]
I just encountered this as I was using the template schema from the next-auth documentation with the new authjs/prisma-adapter. There are some slight changes in schema, and once I used the template from the new documentation, it all worked as expected.
New docs: https://authjs.dev/reference/adapter/prisma#prismaadapter
Old docs: https://next-auth.js.org/v3/adapters/prisma

**Reply by Firephoenix25** [reply]
Worked :  )

**Reply by metal-face** [reply]
Thank you very much for this

**Reply by quanzero1992** [reply]
Perfect!!!

**Reply by isabelleilyia** [reply]
where can I see the correct prisma schema in these links?

**Reply by nicoalbanese** [reply]
here: https://authjs.dev/reference/adapter/prisma#create-the-prisma-schema-from-scratch

**Reply by IanAndy** [reply]
This works. Thank you

**Reply by akshatarora130** [reply]
Thanks @nicoalbanese This worked for me as well

**Reply by KeturahDev** [reply]
This saved me as well! Thanks!

**Reply by alexOS-dev** [reply]
It worked for me too, thank you!

**Reply by deepakyadav20322** [reply]
It works for me, thanks.

**Reply by dimerbwimba** [reply]
This works for me thanks @estevanjantsk

**Reply by mustafazuberiexp** [reply]
It worked for me too, thank you!

**Reply by lmy0125** [reply]
This is the solution!

**Comment by 777777Z77777777** [comment]
God bless you!

пт, 8 сент. 2023 г. в 21:10, Nico ***@***.***>:

**Comment by umanda** [comment]
Despite exploring various solutions, the issue persists. For context, I'm working with the following versions in my project:

```
next: 14.1.0
next-auth: ^4.24.6
@next-auth/prisma-adapter: ^1.0.7
@prisma/client: ^5.10.2
```

After encountering this challenge, I decided not to spend much time on it. Instead, I opted for a more flexible approach by implementing a Custom Adapter. This has given me full control over the authentication process. While this workaround suits my current needs, I'm still interested in any insights or solutions that might address the original issue with the prescribed setup.

I appreciate any guidance or feedback from the community.

**Reply by isabelleilyia** [reply]
Do you mind sharing your solution? Would be much appreciated :)

**Reply by umanda** [reply]
Certainly,

In my scenario, I utilise an external API as my backend infrastructure. To integrate this effectively, I devised a bespoke adaptor, detailed as follows in my `my-adaptor.ts` file:

```typescript
import axios from "axios";

export async function apiService(url: string, body: any) {
  try {
    const { data } = await axios.post(
      url,
      body,
      { headers: { "Content-Type": "application/json" } }
    );
    return data;
  } catch (err: any) {
    throw new Error(err.message || "Something went wrong.");
  }
}

/** @return { import("next-auth/adapters").Adapter } */
export function MyAdapter() {
  return {
    async createUser(user: any): Promise<any> {
      const request = await apiService("<API endpoint>", {<data>});
      return request;
    },

    async test(id: any): Promise<any> {
      return;
    },

    async template(updatedUser: any): Promise<any> {
      return;
    },
  };
}

// Configuration for NextAuth options
export const options: NextAuthOptions = {
  adapter: MyAdapter(),
  // ... the remainder of your code
};
```

This custom adaptor is pivotal for the seamless integration with NextAuth, facilitating user creation and other operations by communicating with the external API. It exemplifies a tailored solution for specific backend requirements, ensuring that the application remains flexible and maintainable.


This approach only for my custom need. you may get idea from this.
